R/renamed-files.R

```diff
@@ -0,0 +1,151 @@
+#' Set the column  `changed_files` of a `log` to the latest name of the file
+#'
+#' [parse_log_detailed()] parses a git log. However, changed file names
+#' are not corrected retroactively. Hence, if one wants to process file-specific
+#' commit data for one file across different names for that file, the column
+#' `changed_files` has to be unified. This is done with this function by
+#' updating the file names recursively and setting them to the latest name used.
+#' @param log An unnested detailed log.
+#' @examples
+#' library("magrittr")
+#' gitsumlog %>%
+#'   tidyr::unnest() %>%
+#'   set_changed_file_to_latest_name() %>%
+#'   add_line_history()
+#' @export
+set_changed_file_to_latest_name <- function(log) {
+  is_renaming <- is_name_change(log$changed_file)
+  contains_renaming <- log[is_renaming, ]
+  contains_no_renaming <- log[!is_renaming,]
+  transition_model <- parse_reassignment(
+    contains_renaming$changed_file, which(is_renaming)
+  )
+  update_changed_file_sequentially(log, transition_model)
+}
+
+
+#' Detect whether a `changed_file`-entry is a renaming
+#'
+#' @param changed_file The column `changed_file` from a parsed unnested
+#'   detailed log.
+#' @param reassignment_pattern The regular expression pattern that corresponds
+#'   to renaming.
+#' @importFrom stringr str_locate
+#' @examples
+#' gitsum:::is_name_change("R/{gitsum.R => gitsum-package.R}")
+is_name_change <- function(changed_file,
+                           reassignment_pattern = "\\{.+\\s\\=\\>.+\\}$") {
```

Comment:
I noticed that this pattern might miss some changed files due to truncation of `log$changed_file`.

e.g. see [man/ensure_gitusm_repo.Rd → man/ensure_gitsum_repo.Rd](https://github.com/lorenzwalthert/gitsum/commit/bc424b294eb239a770d683a5dba5f0103668ded5#diff-2f53785f2310dc1682ed4bce1e134c45)

To verify that this get's missed run the below code from within your gitsum repo:

```
unnest_log <- parse_log_detailed(".") %>% tidyr::unnest()
unnest_log[is_name_change(unnest_log$changed_file),"changed_file"]
# A tibble: 10 x 1
...
unnest_log[grepl("=>", unnest_log$changed_file),"changed_file"]
# A tibble: 11 x 1
                                         changed_file
                                                <chr>
 1              man/{get_log.Rd => get_log_simple.Rd}
 2             man/{parse_log.Rd => parse_log_one.Rd}
 3    man/{get_pattern.Rd => get_pattern_multiple.Rd}
 4      man/{get_log_regex.Rd => git_log_detailed.Rd}
 5               man/{rmd_simple.Rd => git_report.Rd}
 6                   R/{gitsum.R => gitsum-package.R}
 7 man/{git_log_detailed.Rd => parse_log_detailed.Rd}
 8     man/{git_log_simple.Rd => parse_log_simple.Rd}
 9 ...ensure_gitusm_repo.Rd => ensure_gitsum_repo.Rd}
10                       man/{log.Rd => gitsumlog.Rd}
11               man/{git_report.Rd => report_git.Rd}
```
Do you think it's necessary to check for `{` and `}`? Maybe this regex could/should be relaxed slightly?

Reply:
Thanks for pointing out to that. I have just pushed a quick fix in #19. It essentially sets the environment variable `$COLUMNS` to a high value before writing the log to a file and hence should avoid truncation of the output. Not sure whether that's the way to go. Also, we should probably check that there is no truncated `changed_file` somehow.

Reply:
I just changed the name of a root directory file and the file name with the reassignment  it does not have curly braces, so you are right, we indeed should not require curly braces.